pkg/monitoring/services/services.go

```diff
@@ -0,0 +1,290 @@
+}
+
+// ListUpstartServices list upstart services via `initctl list`. Returns []SysVService because Upstart is compatible with SysVInit and has the same details
+func ListUpstartServices() ([]SysVService, error) {
+	cmd := exec.Command("initctl",
+		"list",
+	)
+	setPathEnvVar(cmd)
+
+	var outb bytes.Buffer
+	cmd.Stdout = &outb
+	cmd.Stderr = &outb
+	err := cmd.Run()
+	if err != nil {
+		errOutput, _ := ioutil.ReadAll(&outb)
+		return nil, fmt.Errorf("Initctl: %s, %s", err.Error(), string(errOutput))
+	}
+
+	var services []SysVService
+	scanner := bufio.NewScanner(&outb)
+	for scanner.Scan() {
+		parts := strings.Fields(scanner.Text())
+
+		if len(parts) < 2 {
+			continue
+		}
+
+		// filter network interfaces from results e.g.
+		// network-interface-security (network-interface/ip6_vti0) start/running
+		// network-interface (dummy0) start/running
+		if strings.HasPrefix(parts[0], "network-interface") {
+			continue
+		}
+
+		stateParts := strings.Split(parts[1], "/")
+
+		services = append(services,
+			SysVService{UnitFile: parts[0], Status: stateParts[1]},
+		)
+	}
+
+	return services, scanner.Err()
+}
+
+func isSystemd() bool {
+	if _, err := os.Stat("/run/systemd/system"); err == nil {
+		return true
+	}
+	return false
+}
+
+func isUpstart() bool {
+	if _, err := os.Stat("/sbin/upstart-udev-bridge"); err == nil {
+		return true
+	}
+	cmd := exec.Command("initctl", "--version")
+	setPathEnvVar(cmd)
+	if out, err := cmd.Output(); err == nil {
+		if strings.Contains(string(out), "initctl (upstart") {
+			return true
+		}
+	}
+
+	return false
+}
+
+func setPathEnvVar(cmd *exec.Cmd) {
+	// path PATH env var to be able to found right exec location
+	cmd.Env = append(cmd.Env, "PATH="+os.Getenv("PATH"))
+}
+
+// ListServices detect the linux system manager and parse/combine results
+func ListServices() (map[string]interface{}, error) {
+	if runtime.GOOS != "linux" {
+		return nil, ErrorNotImplementedForOS
+	}
+
+	if isSystemd() {
+		services, err := ListSystemdServices()
+		if err != nil {
+			// in case of error lets try to query other
+			log.Errorf("[Services] Systemd appears running but failed to list a services: %s", err.Error())
+		} else {
+			var servicesList []map[string]interface{}
+			for _, service := range services {
+				servicesList = append(servicesList,
+					map[string]interface{}{
+						"name":         service.UnitFile,
+						"load_state":   service.LoadState,
+						"active_state": service.ActiveState,
+						"sub_state":    service.SubState,
+						"description":  service.Description,
+						"manager":      "systemd",
+					})
+			}
+			// if systemd returns results we can return here cause systemd can't work simultaneously with sysvinit/upstart
+			return map[string]interface{}{"list": servicesList}, err
+		}
+	}
+
+	sysVServices, err := ListSysVinitServices()
+	if err != nil {
+		// in case of error lets try to query other
+		log.Errorf("[Services] SysVinit: failed to list a services: %s", err.Error())
```

Comment:
Do we continue in error case?

Reply:
Yes. SysVinit is compatible with Upstart. So we should check both